src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1246,192 @@
 
             empty-beta-graph
-            productions)))
+            productions))
+
+(sc/defn extract-exprs :- schema/NodeExprLookup
+  "Walks the Alpha and Beta graphs and extracts the expressions that will be used in the construction of the final network.
+   The extracted expressions are stored by their key, [<node-id> <field-key>], this allows for the function to be retrieved
+   after it has been compiled.
+
+   Note: The keys of the map returned carry the metadata that can be used during evaluation. This metadata will contain,
+   if available, the compile context, file and ns. This metadata is not stored on the expression itself because it will
+   contain forms that will fail to eval."
+  [beta-graph :- schema/BetaGraph
+   alpha-graph :- [schema/AlphaNode]]
+  (let [backward-edges (:backward-edges beta-graph)
+        handle-expr (fn [exprs s-expr id expr-key metadata]
+                      (assoc exprs
+                        ;; Letting the key carry all of the metadata, this makes the batch evaluation of the expressions
+                        ;; easier because the metadata contains unevalable forms.
+                        (with-meta [id expr-key] (assoc metadata expr-key s-expr))
+                        s-expr))
+        is-root? #(= #{0} (get backward-edges %))
+
+        ;; If extract-exprs ever became a hot spot, this could be changed out to use more java interop.
+        alpha-exprs (reduce (fn [prev alpha-node]
+                              (let [{:keys [id condition env]} alpha-node
+                                    {:keys [type constraints fact-binding args]} condition
+                                    cmeta (meta condition)]
+                                (handle-expr prev
+                                             (with-meta (compile-condition
+                                                          type (first args) constraints
+                                                          fact-binding env)
+                                                        ;; Remove all metadata but file and line number
+                                                        ;; to protect from evaluating usafe metadata
+                                                        ;; See PR 243 for more detailed discussion
+                                                        (select-keys cmeta [:line :file]))
+                                             id
+                                             :alpha-expr
+                                             {:file (or (:file cmeta) *file*)
+                                              :compile-ctx {:condition condition
+                                                            :env env
+                                                            :msg "compiling alpha node"}})))
+                            {}
+                            alpha-graph)
+        exprs (reduce-kv (fn [prev id production-node]
```

Comment:
IMO a more specific name here than "exprs" would be helpful.  This will be a map containing expressions if I'm understanding correctly?